improve error messages by making code start on first line

atari2600tim opened the pull request:
<!-- Keep the title short & concise so anyone non-technical can understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions
Behind the scenes there are things like `code = qsl("function Alias%1() %2\nend").arg(id, code)` which handles the code of an alias and embeds it into a function.  Timers, triggers, and keybinds all have similar, but with the `function X()` on a line alone.  When they get compiled, the first line of an alias code box is also the first line of the resulting function, so an error on the first line of the edit box code is also the first line of the post-embed code.  But for the other type of items, the first line of edit box is second line of post-embed code.  When compiling, the system is looking at the post-embed code, so it gives errors with line numbers that make sense with aliases but are confusing with the others.

This edit should make them all similar to aliases.

#### Motivation for adding to Mudlet
fix #1460 
#### Other info (issues closed, discussion etc)
Error messages say things like "near X" instead of citing column numbers, so I don't think there are any negative consequences for first line being on same line after `function X()`.  It's how aliases are handled already.

atari2600tim committed:
make code start on first line

SlySven commented on src/TAction.cpp:
```diff
@@ -121,7 +121,7 @@ bool TAction::setScript(const QString& script)
 {
     mFuncName = QString("Action") + QString::number(mID);
-    const QString code = QString("function ") + mFuncName + QString("()\n") + mScript + QString("\nend\n");
+    const QString code = QString("function %1() %2\nend").arg(mFuncName, mScript);
```
```suggestion
    const QString code = qsl("function %1() %2 end").arg(mFuncName, mScript);
```

atari2600tim replied:
Initially I did this same thing (the `\n`) and then found that the last line of the edited box having a comment would comment out the end.  I don't like how an unclosed thing would say error found on line 11 if there are 10 lines in the edit box, but this may be a necessary evil.

![image](https://github.com/Mudlet/Mudlet/assets/29287358/82943520-505c-4abc-a341-93425ec84ff1)

SlySven replied:
Oh, yeah, I see what you mean - so that isn't going to work. *\*sigh\**

BTW the end bit of that error message `... near "`  - is that a manifestation of the problem with un-escaped HTML symbols in the Lua error message and it should read `... near <eof>"`?

atari2600tim replied:
I did not notice that when I took the screenshot.  You are exactly right, it says `near <eof>`.  I think I did not touch that in the current html escaping PR because I failed to produce such an error message in that spot.  ~~I put a random `<` at the end of some code and it said `unexpected symbol` without specifying the symbol, and eventually I gave up at trying to produce an error message with a `<` in it.  Thanks very much.  I'll add those to it.~~

edit: Well, maybe it is already good.

SlySven commented on src/TKey.cpp:
```diff
@@ -166,7 +166,7 @@ bool TKey::setScript(const QString& script)
 {
     mFuncName = QString("Key") + QString::number(mID);
-    const QString code = QString("function ") + mFuncName + QString("()\n") + mScript + QString("\nend\n");
+    const QString code = QString("function %1() %2\nend").arg(mFuncName, mScript);
```
```suggestion
    const QString code = qsl("function %1() %2 end").arg(mFuncName, mScript);
```

SlySven commented on src/TAction.cpp:
```diff
@@ -121,7 +121,7 @@ bool TAction::setScript(const QString& script)
 bool TAction::compileScript()
 {
     mFuncName = QString("Action") + QString::number(mID);
```
```suggestion
    mFuncName = qsl("Action%1").arg(QString::number(mID));
```

SlySven commented on src/TAction.cpp:
```diff
@@ -121,7 +121,7 @@ bool TAction::setScript(const QString& script)
-    const QString code = QString("function ") + mFuncName + QString("()\n") + mScript + QString("\nend\n");
+    const QString code = QString("function %1() %2\nend").arg(mFuncName, mScript);
     QString error;
     if (mpHost->mLuaInterpreter.compile(code, error, QString("Button: ") + getName())) {
```
```suggestion
    if (mpHost->mLuaInterpreter.compile(code, error, qsl("Button: %1").arg(getName()))) {
```

SlySven commented on src/TKey.cpp:
```diff
@@ -166,7 +166,7 @@ bool TKey::setScript(const QString& script)
 bool TKey::compileScript()
 {
     mFuncName = QString("Key") + QString::number(mID);
```
```suggestion
    mFuncName = qsl("Key%1").arg(QString::number(mID));
```

SlySven commented on src/TKey.cpp:
```diff
@@ -166,7 +166,7 @@ bool TKey::setScript(const QString& script)
-    const QString code = QString("function ") + mFuncName + QString("()\n") + mScript + QString("\nend\n");
+    const QString code = QString("function %1() %2\nend").arg(mFuncName, mScript);
     QString error;
     if (mpHost->mLuaInterpreter.compile(code, error, QString("Key: ") + getName())) {
```
```suggestion
    if (mpHost->mLuaInterpreter.compile(code, error, qsl("Key: %1").arg(getName()))) {
```

SlySven commented on src/TTimer.cpp:
```diff
@@ -193,7 +193,7 @@ bool TTimer::setScript(const QString& script)
 bool TTimer::compileScript()
 {
     mFuncName = QString("Timer") + QString::number(mID);
```
```suggestion
    mFuncName = qsl("Timer%1").arg(QString::number(mID));
```

SlySven commented on src/TTimer.cpp:
```diff
@@ -193,7 +193,7 @@ bool TTimer::setScript(const QString& script)
 {
     mFuncName = QString("Timer") + QString::number(mID);
-    const QString code = QString("function ") + mFuncName + QString("()\n") + mScript + QString("\nend\n");
+    const QString code = QString("function %1() %2\nend").arg(mFuncName, mScript);
```
```suggestion
    const QString code = qsl("function %1() %2 end").arg(mFuncName, mScript);
```

SlySven commented on src/TTimer.cpp:
```diff
@@ -193,7 +193,7 @@ bool TTimer::setScript(const QString& script)
-    const QString code = QString("function ") + mFuncName + QString("()\n") + mScript + QString("\nend\n");
+    const QString code = QString("function %1() %2\nend").arg(mFuncName, mScript);
     QString error;
     if (mpHost->mLuaInterpreter.compile(code, error, "Timer: " + getName())) {
```
```suggestion
    if (mpHost->mLuaInterpreter.compile(code, error, qsl("Timer: %1).arg(getName()))) {
```

SlySven commented on src/TTrigger.cpp:
```diff
@@ -240,7 +240,7 @@ bool TTrigger::setRegexCodeList(QStringList patterns, QList<int> patternKinds)
             func << "trigger" << mID << "condition" << i;
             funcName = func.str();
-            const QString code = qsl("function %1()\n%2\nend\n").arg(funcName.c_str(), patterns[i]);
+            const QString code = qsl("function %1() %2\nend").arg(funcName.c_str(), patterns[i]);
```
```suggestion
            const QString code = qsl("function %1() %2 end").arg(funcName.c_str(), patterns[i]);
```

SlySven replied:
Yeah that suggestion is no good.

SlySven commented on src/TTrigger.cpp:
```diff
@@ -1324,7 +1324,7 @@ bool TTrigger::setScript(const QString& script)
 {
     mFuncName = qsl("Trigger%1").arg(QString::number(mID));
-    const QString code = qsl("function %1()\n%2\nend\n").arg(mFuncName, mScript);
+    const QString code = qsl("function %1() %2\nend").arg(mFuncName, mScript);
```
```suggestion
    const QString code = qsl("function %1() %2 end").arg(mFuncName, mScript);
```

SlySven replied:
No good.

SlySven reviewed (commented):
As I mentioned in another PR I've also been poking under this stone, and in this case I agree with your analysis - however you can go further and also eliminate the line feed inserted at the end of the code block so that references that the Lua interpreted might make to `<eof>` (which will soon be handled better by one of your current PRs if it hasn't already been handled! :grinning: ) might also refer to a sensible line number.

Also, there are a number of `QString(...)` around this neck of the ~~woods~~ *code-base* that ought to wrapped up in something that marks them as NOT to be translated - and in this context the `qsl(`...`)` which is our own version of `QStringLiteral(`...`)` is the most suitable candidate.

You may want to check the code around the same places for the `TAlias` and `TScript` classes...

atari2600tim committed:
suggested changes

QString -> qsl

"Button: "+name  ->  "Button: %1".arg(name)

"Key"+string(number) -> "Key%1".arg(string(number))

atari2600tim committed:
Update TTimer.cpp

atari2600tim committed:
Update src/TKey.cpp

Co-authored-by: Stephen Lyons <slysven@virginmedia.com>

SlySven reviewed (approved):
Okay, LGTM.

atari2600tim committed:
Merge branch 'development' into lineNumbers

atari2600tim committed:
Merge branch 'development' into lineNumbers

vadi2 reviewed (commented):
It looks like we're closer, but the count is still off by 1, right?

Either way a good code improvement :+1:
